src/main/java/io/temporal/internal/common/LocalActivityMarkerData.java

```diff
@@ -95,120 +98,115 @@ public Builder setBackoff(Duration backoff) {
     }
   }
 
-  private static class LocalActivityMarkerHeader {
-    private final String activityId;
-    private final String activityType;
-    private final String errReason;
-    private final long replayTimeMillis;
-    private final int attempt;
-    private final Duration backoff;
-    private final boolean isCancelled;
+  private static class DataValue {
+    private String activityId;
+    private String activityType;
+    private long replayTimeMillis;
+    private int attempt;
+    private long backoffMillis;
+
+    // Needed by Jackson deserializer
+    DataValue() {}
 
-    LocalActivityMarkerHeader(
+    DataValue(
         String activityId,
         String activityType,
         long replayTimeMillis,
-        String errReason,
         int attempt,
-        Duration backoff,
-        boolean isCancelled) {
+        Duration backoff) {
       this.activityId = activityId;
       this.activityType = activityType;
       this.replayTimeMillis = replayTimeMillis;
-      this.errReason = errReason;
       this.attempt = attempt;
-      this.backoff = backoff;
-      this.isCancelled = isCancelled;
+      this.backoffMillis = backoff == null ? 0 : backoff.toMillis();
     }
   }
 
-  private final LocalActivityMarkerHeader headers;
+  private final DataValue data;
   private final Optional<Payloads> result;
+  private final Optional<Failure> failure;
 
   private LocalActivityMarkerData(
       String activityId,
       String activityType,
       long replayTimeMillis,
       Optional<Payloads> result,
-      String errReason,
+      Optional<Failure> failure,
       int attempt,
-      Duration backoff,
-      boolean isCancelled) {
-    this.headers =
-        new LocalActivityMarkerHeader(
-            activityId, activityType, replayTimeMillis, errReason, attempt, backoff, isCancelled);
+      Duration backoff) {
+    this.data = new DataValue(activityId, activityType, replayTimeMillis, attempt, backoff);
     this.result = result;
+    this.failure = failure;
   }
 
-  private LocalActivityMarkerData(LocalActivityMarkerHeader headers, Optional<Payloads> result) {
-    this.headers = headers;
+  private LocalActivityMarkerData(
+      DataValue data, Optional<Payloads> result, Optional<Failure> failure) {
+    this.data = data;
     this.result = result;
+    this.failure = failure;
   }
 
   public String getActivityId() {
-    return headers.activityId;
+    return data.activityId;
   }
 
   public String getActivityType() {
-    return headers.activityType;
-  }
-
-  public String getErrReason() {
-    return headers.errReason;
+    return data.activityType;
   }
 
-  public Optional<Payloads> getErrJson() {
-    return Strings.isNullOrEmpty(headers.errReason) ? Optional.empty() : result;
+  public Optional<Failure> getFailure() {
+    return failure;
   }
 
   public Optional<Payloads> getResult() {
     return result;
   }
 
   public long getReplayTimeMillis() {
```

Comment:
Mixed time types? ReplayTimeMillis vs Backoff

Reply:
ReplayTimeMillis is a specific time of replay (timestamp). Backoff is a duration which is time interval.